// @the-via/reader/dist/transform.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLightingDefinition = exports.keyboardDefinitionV2ToVIADefinitionV2 = exports.keyboardDefinitionV3ToVIADefinitionV3 = exports.getVendorProductId = void 0;
var keyboard_definition_v3_validator_1 = __importDefault(require("./validated-types/keyboard-definition-v3.validator"));
var keyboard_definition_v2_validator_1 = __importDefault(require("./validated-types/keyboard-definition-v2.validator"));
var lighting_presets_1 = require("./lighting-presets");
var validate_1 = require("./validate");
var getHexHint = function (value) {
    var borkedHexPattern = /^[Oo]x/;
    return value.match(borkedHexPattern)
        ? "Did you mean '".concat(value.replace(borkedHexPattern, '0x'), "' instead?")
        : '';
};
var getVendorProductId = function (_a) {
    var productId = _a.productId, vendorId = _a.vendorId;
    if (vendorId.toUpperCase() === '0XFEED') {
        throw new Error("'0xFEED' is not a valid vendorId.");
    }
    var parsedVendorId = parseInt(vendorId, 16);
    var parsedProductId = parseInt(productId, 16);
    if (isNaN(parsedVendorId)) {
        throw new Error("vendorId could not be parsed: '".concat(vendorId, "'. ").concat(getHexHint(vendorId)));
    }
    if (isNaN(parsedProductId)) {
        throw new Error("productId could not be parsed: '".concat(productId, "'. ").concat(getHexHint(productId)));
    }
    return parsedVendorId * 65536 + parsedProductId;
};
exports.getVendorProductId = getVendorProductId;
var keyboardDefinitionV3ToVIADefinitionV3 = function (definition) {
    var _a = (0, keyboard_definition_v3_validator_1.default)(definition), name = _a.name, menus = _a.menus, keycodes = _a.keycodes, customKeycodes = _a.customKeycodes, matrix = _a.matrix, layouts = _a.layouts, firmwareVersion = _a.firmwareVersion;
    var viaLayout = (0, validate_1.validateLayouts)(layouts);
    var keymap = layouts.keymap, partialLayout = __rest(layouts, ["keymap"]);
    var viaLayouts = __assign(__assign({}, partialLayout), viaLayout);
    (0, validate_1.validateKeyBounds)(matrix, viaLayouts);
    (0, validate_1.validateCommonMenus)(menus !== null && menus !== void 0 ? menus : []);
    return {
        name: name,
        vendorProductId: (0, exports.getVendorProductId)(definition),
        firmwareVersion: firmwareVersion !== null && firmwareVersion !== void 0 ? firmwareVersion : 0,
        menus: menus !== null && menus !== void 0 ? menus : [],
        keycodes: keycodes !== null && keycodes !== void 0 ? keycodes : [],
        customKeycodes: customKeycodes,
        matrix: matrix,
        layouts: viaLayouts,
    };
};
exports.keyboardDefinitionV3ToVIADefinitionV3 = keyboardDefinitionV3ToVIADefinitionV3;
var keyboardDefinitionV2ToVIADefinitionV2 = function (definition) {
    var _a = (0, keyboard_definition_v2_validator_1.default)(definition), name = _a.name, customFeatures = _a.customFeatures, customMenus = _a.customMenus, customKeycodes = _a.customKeycodes, lighting = _a.lighting, matrix = _a.matrix, layouts = _a.layouts;
    var viaLayout = (0, validate_1.validateLayouts)(layouts);
    var keymap = layouts.keymap, partialLayout = __rest(layouts, ["keymap"]);
    var viaLayouts = __assign(__assign({}, partialLayout), viaLayout);
    (0, validate_1.validateKeyBounds)(matrix, viaLayouts);
    return {
        name: name,
        lighting: lighting,
        layouts: viaLayouts,
        matrix: matrix,
        customFeatures: customFeatures,
        customKeycodes: customKeycodes,
        customMenus: customMenus,
        vendorProductId: (0, exports.getVendorProductId)(definition),
    };
};
exports.keyboardDefinitionV2ToVIADefinitionV2 = keyboardDefinitionV2ToVIADefinitionV2;
var getLightingDefinition = function (definition) {
    return typeof definition === 'string'
        ? lighting_presets_1.LightingPreset[definition]
        : __assign(__assign({}, lighting_presets_1.LightingPreset[definition.extends]), definition);
};
exports.getLightingDefinition = getLightingDefinition;
